### PR TITLE
[LS] Add completion support for mapping match pattern

### LINKED ISSUE #33383: [LS] Record fields are not suggested in mapping match pattern
**Description:**
Consider the following cursor position. 

```ballerina
public type MyRecord record {
      int field1;
      int field2;
};

const temp = 1;
public function main() {
    MyRecord r1= {field1: 0, field2: 0};

    int myInt = 1;

    match r1 {
          {field1:0, <cursor>} => {
                
          }
    }
}
```

Currently, auto-completions are not provided for record fields in the mapping match pattern context. 
mapping_pattern

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
$subject as depicted below. 

capture

Fixes #33383 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by mohanvive — APPROVED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+        if (evalNodeAtCursor.kind() == SyntaxKind.SPECIFIC_FIELD) {
+            colon = ((SpecificFieldNode) evalNodeAtCursor).colon().orElse(null);
+        } else if (evalNodeAtCursor.kind() == SyntaxKind.COMPUTED_NAME_FIELD) {
+            colon = ((ComputedNameFieldNode) evalNodeAtCursor).colonToken();
+        }
+
+        if (colon == null) {
+            return false;
+        }
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:69]: just return at the else clause of the previous if-else ladder

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            completionItems.addAll(this.getCompletionItemList(variables, context));
+        }
+        return completionItems;
+    }
+
+    protected List<LSCompletionItem> getCompletionsInValueExpressionContext(BallerinaCompletionContext context) {
+        if (QNameReferenceUtil.onQualifiedNameIdentifier(context, context.getNodeAtCursor())) {
+            QualifiedNameReferenceNode qNameRef = (QualifiedNameReferenceNode) context.getNodeAtCursor();
+            return this.getExpressionsCompletionsForQNameRef(context, qNameRef);
+        } else {
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:176]: We can drop the else clause and get the return outside of the else

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingMatchPatternNodeContext.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Override
+    protected boolean withinValueExpression(BallerinaCompletionContext context, NonTerminalNode evalNodeAtCursor) {
+        Token colon = null;
+        if (evalNodeAtCursor.kind() == SyntaxKind.FIELD_MATCH_PATTERN) {
+            colon = ((FieldMatchPatternNode) evalNodeAtCursor).colonToken();
+        }
+        if (colon == null) {
+            return false;
+        }
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingMatchPatternNodeContext.java:89]: Merge with the previous if clause as its else clause

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java (hunk 1 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+/**
+ * Abstract completion provider for mapping constructor context.
+ *
+ * @param <T> mapping constructor node type.
+ * @since 2.0.0
+ */
+public abstract class MappingContextProvider<T extends Node> extends AbstractCompletionProvider<T> {
+
+    protected abstract Map<String, RecordFieldSymbol> getValidFields(T node,
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:62]: This abstract method should come after (Ideally after the other protected methods. But that's opinion based. So not a hard requirement) the constructor.

**Review by IMS94 — CHANGES_REQUESTED** [review]
Implementation looks good. Need to do an ordering change of methods.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java (hunk 4 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java]: Done

**Review by nadeeshaan — DISMISSED** [review]
LGTM

**Review by mohanvive — APPROVED** [review]

**Comment by mohanvive** [comment]
@IMS94 Need your approval.

**Review by IMS94 — APPROVED** [review]
@mohanvive Done!

**Pull request merged**